packages/gatsby/src/commands/develop.ts

```diff
@@ -1,671 +1,66 @@
-      prettyHost = `localhost`
-
-      try {
-        // This can only return an IPv4 address
-        lanUrlForConfig = address.ip()
-        if (lanUrlForConfig) {
-          // Check if the address is a private ip
-          // https://en.wikipedia.org/wiki/Private_network#Private_IPv4_address_spaces
-          if (
-            /^10[.]|^172[.](1[6-9]|2[0-9]|3[0-1])[.]|^192[.]168[.]/.test(
-              lanUrlForConfig
-            )
-          ) {
-            // Address is private, format it for later use
-            lanUrlForTerminal = prettyPrintUrl(lanUrlForConfig)
-          } else {
-            // Address is not private, so we will discard it
-            lanUrlForConfig = undefined
-          }
-        }
-      } catch (_e) {
-        // ignored
-      }
-    }
-    // TODO collect errors (GraphQL + Webpack) in Redux so we
-    // can clear terminal and print them out on every compile.
-    // Borrow pretty printing code from webpack plugin.
-    const localUrlForTerminal = prettyPrintUrl(prettyHost)
-    const localUrlForBrowser = formatUrl(prettyHost)
-    return {
-      lanUrlForConfig,
-      lanUrlForTerminal,
-      localUrlForTerminal,
-      localUrlForBrowser,
-    }
-  }
-
-  function printInstructions(appName: string, urls: IPreparedUrls): void {
-    console.log()
-    console.log(`You can now view ${chalk.bold(appName)} in the browser.`)
-    console.log()
-
-    if (urls.lanUrlForTerminal) {
-      console.log(
-        `  ${chalk.bold(`Local:`)}            ${urls.localUrlForTerminal}`
-      )
-      console.log(
-        `  ${chalk.bold(`On Your Network:`)}  ${urls.lanUrlForTerminal}`
-      )
-    } else {
-      console.log(`  ${urls.localUrlForTerminal}`)
-    }
+  // Run the actual develop server on a random port, and the proxy on the program port
+  // which users will access
+  const userPort = program.port
+  const randomPort = await getPort()
+  console.log({ userPort, randomPort })
+  const script = createControllableScript(report.stripIndent`
+    const cmd = require("${developProcessPath}");
+    const args = ${JSON.stringify({
+      ...program,
+      port: randomPort,
+    })};
+    cmd(args);
+  `)
+
+  startDevelopProxy({
+    proxyPort: userPort,
+    targetPort: randomPort,
+  })
 
-    console.log()
-    console.log(
-      `View ${
-        process.env.GATSBY_GRAPHQL_IDE === `playground`
-          ? `the GraphQL Playground`
-          : `GraphiQL`
-      }, an in-browser IDE, to explore your site's data and schema`
-    )
-    console.log()
+  script.start()
 
-    if (urls.lanUrlForTerminal) {
-      console.log(
-        `  ${chalk.bold(`Local:`)}            ${
-          urls.localUrlForTerminal
-        }___graphql`
-      )
-      console.log(
-        `  ${chalk.bold(`On Your Network:`)}  ${
-          urls.lanUrlForTerminal
-        }___graphql`
+  chokidar
+    .watch([rootFile(`gatsby-config.js`), rootFile(`gatsby-node.js`)])
+    .on(`change`, filePath => {
```

Comment:
I know it's a draft so this is more pre-emptive (so I don't forgot about this later on) than actual comment that need to be handled soon,
but there are some gatsby-config edits that do not require restarts (like editing `siteMetadata` field, there might be more).

So there will be need to add some conditions (either when to restart or when to not restart)